core/src/main/java/org/aerogear/mobile/core/MobileCore.java

```diff
@@ -2,273 +2,174 @@
-        }
+    public static MobileCore init(Context context) throws InitializationException {
+        return init(context, new Options());
     }
 
     /**
-     * This method sorts a list of servers into the order they will need to be initialized in.
+     * Initialize the AeroGear system
      *
-     * @param declaredServices a list of services to be sorted into the order they will be
-     *                         initialized in.
-     * @return a sorted list of services.
-     * @throws BootstrapException if circular or undefined dependencies are detected.
+     * @param context Application context
+     * @param options AeroGear initialization options
      */
-    private List<String> sortServicesIntoBootstrapOrder(List<String> declaredServices) {
-        List<String> workingDeclaredServicesList = new ArrayList<>(declaredServices);
-        List<String> sortedServices = new ArrayList<>(declaredServices.size());
+    public static MobileCore init(Context context, Options options) throws InitializationException {
+        return new MobileCore(context, options);
+    }
 
-        while (!workingDeclaredServicesList.isEmpty()) {
-            sortedServices.add(popNextService(workingDeclaredServicesList, sortedServices));
+    /**
+     * Called when mobile core instance need to be destroyed
+     */
+    public void destroy() {
+        for (Class<? extends ServiceModule> serviceKey : services.keySet()) {
+            ServiceModule serviceModule = services.get(serviceKey);
+            serviceModule.destroy();
         }
+    }
 
-        return sortedServices;
+    public ServiceModule getInstance(Class<? extends ServiceModule> serviceClass) {
+        return getInstance(serviceClass, null);
     }
 
-    /**
-     * Removes a service from the list that can be instanciated or has all of its dependencies in
-     * sortedServices.
-     *
-     * @param workingList    a mutable list to find the first element in that can be resolved given
-     *                       the values in sortedServices
-     * @param sortedServices services which have had their dependencies check and are in initialization
-     *                       order
-     * @return the next service name
-     * @throws BootstrapException if circular or undefined dependencies are detected.
-     */
-    private String popNextService(List<String> workingList, List<String> sortedServices) {
-        boolean unresolvableServiceDetected = true;
-        String serviceUnderInstanciation = "";
-
-        for (String serviceName : workingList) {
-            List<String> dependencies = serviceRegistry.getDependenciesFor(serviceName);
-            serviceUnderInstanciation = serviceName;
-            if (dependencies.isEmpty() || sortedServices.containsAll(dependencies)) {
-                unresolvableServiceDetected = false;
-                break;
-            }
+    public ServiceModule getInstance(Class<? extends ServiceModule> serviceClass,
```

Comment:
How do we resolve concrete vs abstract super types?

For instance if we have a OkHTTPModule in services, but someone calls getInstance(HttpModule.class) this will fail because HttpModule is a interface.